### ISSUE TITLE
Virtual IP issues with keepalived

### ISSUE THREAD

**Issue by sajusat** [vitabaks/postgresql_cluster#109]
There is an issue with the keepalived. For some reason I had the virual IP was not accessible (probably due to a network issue). 
I think what happened was a communication break between MASTER (1st node) and BACKUP vm's and one of the BACKUP machine has send out G-ARP message and the ARP entry was changed to the MAC adderss of the BACKUP (2nd node).  I restarted the keepalived service on MASTER (1st node) and it fixed the issue of virtual IP being inaccessible. Now I looked up all the nodes and all the keepalived.conf are in BACKUP state.  The log on the 2nd node says it is in the MASTER state (This a sync standby machine and the MASTER is acutally the postgres leader machine).

This is log from the 2nd node 
```
Jul 24 17:00:44 prod-pg2 Keepalived[32565]: Starting Keepalived v1.3.9 (10/21,2017)
Jul 24 17:00:44 prod-pg2 Keepalived[32565]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 24 17:00:44 prod-pg2 systemd[1]: Started Keepalive Daemon (LVS and VRRP).
Jul 24 17:00:44 prod-pg2 Keepalived[32567]: Starting Healthcheck child process, pid=32571
Jul 24 17:00:44 prod-pg2 Keepalived_healthcheckers[32571]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 24 17:00:44 prod-pg2 Keepalived[32567]: Starting VRRP child process, pid=32573
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: Registering Kernel netlink reflector
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: Registering Kernel netlink command channel
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: Registering gratuitous ARP shared channel
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: Using LinkWatch kernel netlink reflector...
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: VRRP_Instance(VI_1) Entering BACKUP STATE
Jul 24 17:00:44 prod-pg2 Keepalived_vrrp[32573]: VRRP_Script(haproxy_check) succeeded
Jul 24 17:00:46 prod-pg2 Keepalived_vrrp[32573]: VRRP_Instance(VI_1) Changing effective priority from 100 to 102
Jul 24 17:00:50 prod-pg2 Keepalived_vrrp[32573]: VRRP_Instance(VI_1) Transition to MASTER STATE
Jul 24 17:00:52 prod-pg2 Keepalived_vrrp[32573]: VRRP_Instance(VI_1) Entering MASTER STATE
```

This is the log from the current postgres leader (previously MASTER)
```
Jul 24 17:00:36 prod-pg1 systemd[1]: keepalived.service: Found left-over process 1554 (keepalived) in control group while starting unit. Ignoring.
Jul 24 17:00:36 prod-pg1 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Jul 24 17:00:36 prod-pg1 systemd[1]: keepalived.service: Found left-over process 1557 (keepalived) in control group while starting unit. Ignoring.
Jul 24 17:00:36 prod-pg1 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Jul 24 17:00:36 prod-pg1 systemd[1]: keepalived.service: Found left-over process 1559 (keepalived) in control group while starting unit. Ignoring.
Jul 24 17:00:36 prod-pg1 systemd[1]: This usually indicates unclean termination of a previous run, or service implementation deficiencies.
Jul 24 17:00:36 prod-pg1 systemd[1]: Starting Keepalive Daemon (LVS and VRRP)...
Jul 24 17:00:37 prod-pg1 Keepalived[22617]: Starting Keepalived v1.3.9 (10/21,2017)
Jul 24 17:00:37 prod-pg1 Keepalived[22617]: Opening file '/etc/keepalived/keepalived.conf'.
Jul 24 17:00:37 prod-pg1 Keepalived[22617]: daemon is already running
Jul 24 17:00:37 prod-pg1 systemd[1]: Started Keepalive Daemon (LVS and VRRP).
Jul 26 04:30:08 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) timed out
Jul 26 04:30:08 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 102 to 100
Jul 26 04:30:08 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:11 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:13 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:15 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:17 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:19 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:20 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) succeeded
Jul 26 04:30:22 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 100 to 102
Jul 26 04:30:23 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) timed out
Jul 26 04:30:23 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:24 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 102 to 100
Jul 26 04:30:25 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:27 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:28 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) succeeded
Jul 26 04:30:28 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 100 to 102
Jul 26 04:30:32 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) timed out
Jul 26 04:30:32 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:33 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 102 to 100
Jul 26 04:30:33 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:34 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) succeeded
Jul 26 04:30:35 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 100 to 102
Jul 26 04:30:37 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) timed out
Jul 26 04:30:37 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 102 to 100
Jul 26 04:30:37 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:40 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:30:41 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) succeeded
Jul 26 04:30:41 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 100 to 102
Jul 26 04:31:24 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) timed out
Jul 26 04:31:25 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 102 to 100
Jul 26 04:31:26 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:31:29 prod-pg1 Keepalived_vrrp[1559]: /usr/libexec/keepalived/haproxy_check.sh exited due to signal 15
Jul 26 04:31:30 prod-pg1 Keepalived_vrrp[1559]: VRRP_Script(haproxy_check) succeeded
Jul 26 04:31:31 prod-pg1 Keepalived_vrrp[1559]: VRRP_Instance(VI_1) Changing effective priority from 100 to 102
```
I believe there is a zombie process from the previous machine restart.

Now both machines have the priority 102 as per the logs, but the configuration on each machine is the same.

``` 
global_defs {
   router_id ocp_vrrp
   enable_script_security
   script_user root
}

vrrp_script haproxy_check {
   script "/usr/libexec/keepalived/haproxy_check.sh"
   interval 2
   weight 2
}

vrrp_instance VI_1 {
   interface ens3
   virtual_router_id 201
   priority  100
   advert_int 2
   state  BACKUP
   virtual_ipaddress {
       10.200.162.201
   }
   track_script {
       haproxy_check
   }
   authentication {
      auth_type PASS
      auth_pass xxxxxx
   }
}
```
What is the correct way to restart keepalived to set the 1st node to MASTER state. 

should I add the `garp_master_refresh` configuration to the vrrp_instance VI_1 code block so that the issues do not happen again if the communication breaks for some reason between the current MASTER and BACKUP nodes ?

**Comment by tyldum** [comment]
+1
I am yet to dissect an issue we had last week that seems identical. I am om vacation so I haven't had time to dive into it, but a restart of keepalived solved it.

**Comment by tyldum** [comment]
My issue was traced down to this Ubuntu bug after an systemd upgrade: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1815101

**Comment by sajusat** [comment]
Yes, the problem can overcome by selectively restarting the keepalived services. 
@tyldum I will check the bug report and see if there is any updates available to for it. I am also using Ubuntu (18.04).
@vitabaks The default run of ansible playbook install a very old version of keepalived. Is there any particular reason for that ?

**Comment by vitabaks** [comment]
>  The default run of ansible playbook install a very old version of keepalived. 

This playbook actually installs the version of the keepalived package that is located in the repository of this Linux distribution.
Specify which version (and why) we need to install

**Comment by sajusat** [comment]
@vitabaks I was just following the bug thread that @tyldum  reported. It says that the Keepalived versions > 2.0. handles this correctly. 
Moreover, when I was looking for the possible causes; one of the reasons that came up was the `garp_master_refresh` messages. It is only supported in the later versions of keepalived. In my case the keepalived version is 1.3.9.

**Comment by tyldum** [comment]
A patched systemd for Ubuntu Bionic was released this week and addresses
the problem. Be aware that the upgrade itself will trigger the problem one
last time (it did when I applied it at least).

Ubuntu Focal had a fix earlier.

Upgrade and see if you observe the same issue, if so you have confirmed the
root cause and hopefully it will be the end of it.

It broke a lot of ha-type services on Ubuntu, not just keepalived.

I don't belive the problem is related to this repo and the keepalived
config.

tor. 19. aug. 2021, 17:00 skrev Saju ***@***.***>:
